### PR TITLE
DB-9455 Refactor TimestampOracle to expose three distinct functions

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:92]: **synchronized** does not address the issue described in the bug.

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:112]: Recursion is not justified. A simple loop is clearer and can be used in the calling method:
`while (nextTimestamp  > _maxReservedTimestamp ) {
    reserveNextBlock(_maxReservedTimestamp);
}
`
I wouldn't worry about "double volatile read": it's not on the hot path and synchronization cost along with ZK read make its extra cost negligible.

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:100]: nextTS ?

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:113]: If **nextTS** is larger than **_maxReservedTimestamp** by several **blockSize**, we'll go to ZK several times.
Why do that when we can do the right thing in one take?

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:104]: Double volatile read is gone. Can get rid of maxTS.

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:97]: By calling **reserveNextBlock(timestamp)** we have guaranteed that **timestamp <= _maxReservedTimestamp**. Reassigning **_maxReservedTimestamp** to  **timestamp** doesn't look right.

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:98]: **synchronized** doesn't guard **_timestampCounter** . I'd still like to see **compareAndSet** here.

File: splice_timestamp_api/pom.xml
Comment (arnaud-splice) [reply splice_timestamp_api/pom.xml:46]: done

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:104]: Done

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:113]: Fixed

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:98]: done

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:97]: I'm not quite sure what else to put instead. @jyuanca, can you weigh in?

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:112]: blockCount is not computed right (can be zero or even negative).

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:117]: Use _addAndGet()_

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:97]: _reserveNextBlock_ assigns a new value to _maxReservedTimestamp, here it's overwritten.
updateAndGet ensures assignment (it has a loop), which is more than needed here; a simple _compareAndSet_ with the value from the initial _timestampCounter.get() would suffice.

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:112]: Fixed

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:117]: Done

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (jyuanca) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:113]: Change to nextTS  - _maxReservedTimestamp to make it positive? Just in case next Ts is much larger than _maxReservedTimestamp and blockCount becomes negative.

Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java]: Done

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java]: Done

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1408/

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:126]: "while" is redundant here, "if" would suffice

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (OlegMazurov) [inline splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java:111]: _compareAndSet_ returns _true_ if successful, which also guarantees the condition is satisfied.
Consider:

	public synchronized void bumpTimestamp(long newTimestamp) throws TimestampIOException {
		newTimestamp = roundUp(newTimestamp);
		if (newTimestamp> _maxReservedTimestamp) {
			reserveNextBlock(newTimestamp);
		}
		long currentTimestampCounter;
		while ((currentTimestampCounter = _timestampCounter.get()) < newTimestamp) {
			if (_timestampCounter.compareAndSet(currentTimestampCounter, newTimestamp)) break;
		}
	}

**Comment by arnaud-splice** [comment]
The error in IT seems to be the one that DB-10025 fixes. I'll wait until it's merged and will trigger ITs again.

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java]: done

File: splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java
Comment (arnaud-splice) [reply splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java]: done

**Review by dgomezferro — APPROVED** [review]
Nice cleanup!

**Pull request merged**
